### DISCUSSION TITLE
Decoding event logs

### DISCUSSION THREAD

**Discussion by mik-0** [wevm/viem#501] [Question] [answered]
Hi there,

I recently started using Viem and so far it has been great! Unfortunately I'm running into some problems trying to decode event logs. Currently, my code looks like this (I want to extract the uniqueId argument from the event):

```ts
const unwatch = client.watchContractEvent({
    address: config.DAO_CONTRACT_ADDRESS as any,
    abi,
    eventName: "PaymentProcessed",
    onLogs: (logs) => {
        for (const log of logs) {
            const topics = decodeEventLog({
                abi,
                data: log.data,
                topics: ... // ?
            });
            const uniqueId = ... // ?
        }
    },
});
```

As you can see, I'm not sure what I should put in the topics field. I tried using encodeEventTopics but I could not get the types to match. I could not find a complete example for decoding contract events. I'm not even sure if I'm doing this correctly.

I would appreciate it if anyone could help me out :)

**Comment by djeelun** [comment]
following

**Comment by jxom** [comment]
The logs should already be decoded on `onLogs` via the `args` attribute on a log. So you shouldn’t need to decode the event log at all.

**Reply by mik-0** [reply]
Thanks for the quick reply! That was what I used originally, but after updating viem it told me the args attribute does not exist on the log object:
<img width="275" alt="image" src="https://github.com/wagmi-dev/viem/assets/23509561/52902943-bb2a-4882-8fb9-b272879204f2">

I'm currently on viem 0.3.21

**Reply by jxom** [reply, marked as answer]
It definitely does exist. Is your ABI cast to `as const`? 

Can you reproduce the type error in the TypeScript playground? https://www.typescriptlang.org/play
(Importing viem should automatically install it)

**Reply by mik-0** [reply]
> It definitely does exist. Is your ABI cast to `as const`?
> 
> Can you reproduce the type error in the TypeScript playground? https://www.typescriptlang.org/play (Importing viem should automatically install it)

Oh my bad, I did not cast the ABI to `as const` 😬. Thanks for the help!

**Reply by jxom** [reply]
One day TypeScript will be smart enough 🤓

**Comment by dsldsl** [comment]
@jxom should this also work for watchEvent?  and should it work when passing a string in to parseAbiItem to the event?  (in this context I can't cast it as const)

I can't seem to get it to parse the logs and provide the args variable in the logs for something like this, though it does correctly filter and pass onLogs the events:

viemPublicClient.watchEvent({
  event: parseAbiItem("event Swap(address indexed sender, uint amount0In, uint amount1In, uint amount0Out, uint amount1Out, address indexed to)"),
  onLogs: (logs) => { 
    logs.forEach((log) => { console.log(JSON.stringify(log))})
  },
  onError: (error) => { console.error("event swap error" + error.message, error) },
});

On a probably-related note, I note that I can't filter on args from watchEvent in this pattern either.
